common/fs/src/cache/dir_path.rs

```diff
@@ -74,3 +75,59 @@ impl From<DirPathBuf> for PathBuf {
         d.inner
     }
 }
+
+// Recursive directory back off
+fn find_valid_path(
+    path: Option<PathBuf>,
+    postfix: Option<PathBuf>,
+) -> Result<DirPathBuf, DirPathBufError> {
+    match path {
+        Some(p) => {
+            if p.is_dir() {
+                Ok(DirPathBuf { inner: p, postfix })
+            } else {
+                warn!("{:?} is not a directory; moving to parent directory", p);
+                let mut postfix_pathbuf = PathBuf::new();
+                let mut pop_pathbuf = PathBuf::new();
+                if let Some(path) = postfix {
+                    pop_pathbuf.push(path);
+                }
+                let parent = level_up(&p).unwrap();
+                let tmp_dir = parent;
+                let postfix_path = p.strip_prefix(tmp_dir).ok();
+                postfix_pathbuf.push(postfix_path.unwrap());
+                postfix_pathbuf.push(pop_pathbuf);
+                find_valid_path(level_up(&p), Some(postfix_pathbuf))
+            }
+        }
+        None => Err(DirPathBufError::NotADirPath(path.unwrap())),
+    }
+}
+
+fn level_up(path: &Path) -> Option<PathBuf> {
+    let mut parent_path = PathBuf::new();
+    match path.parent() {
+        Some(p) => {
+            parent_path.push(p);
+            Some(parent_path)
+        }
+        None => None,
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn test_level_up() {
```

Comment:
find_valid_path has various corner cases...

Reply:
True. In this case the `level_up` function was easy to cover in a simple unit test. There are other unit and integration test that do seem to cover path creation. Are there any specific cases you suggest testing?

Reply:
@dkhokhlov there is also an integration test that was merged before this PR to specifically test, I activated it in the last commit.

Reply:
root, w/o any "/", file, invalid path

Reply:
Added additional code coverage around `find_valid_path` function.